README.rst
==============
bda.plone.cart
==============

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/ambv/black

.. image:: https://travis-ci.org/bluedynamics/bda.plone.shop.svg?branch=master
    :target: https://travis-ci.org/bluedynamics/bda.plone.shop


Shopping cart for plone.

.. contents::


Base package Installation
=========================

Depend your instance to ``bda.plone.cart`` and install it as addon
in plone control panel.


Provide contracts
=================

This package provides components needed to render a shopping cart. It expects
several contracts defined in:

- ``bda.plone.cart.interfaces.ICartDataProvider``
- ``bda.plone.cart.interfaces.ICartItemDataProvider``
- ``bda.plone.cart.interfaces.ICartItemStock``
- ``bda.plone.cart.interfaces.ICartItemPreviewImage``
- ``bda.plone.cart.interfaces.ICartItemAvailability``
- ``bda.plone.cart.interfaces.ICartItemState``
- ``bda.plone.cart.interfaces.ICartDiscount``
- ``bda.plone.cart.interfaces.ICartItemDiscount``

Please take a look at the corresponding interfaces for contract details.
Further some abstract base implementations are available in
``bda.plone.cart.__init__`` which can be used as base class for concrete
implementations.


Cart Visibility
===============

The cart can be rendered as a portlet or inside a viewlet in the portal
header. It's not possible to render the cart twice in one page. Thus, the
viewlet gets skipped automatically if a cart portlet assignment is found.
Also, if cart summary, checkout, order confirmation or portal factory is
rendered, regular cart rendering gets skipped.

Its possible to define custom patterns which are checked against recent
browser URL defining whether to skip cart rendering. This is useful for
custom payment implementations and similar:

.. code-block:: python

    from bda.plone.cart.browser.portlet import SKIP_RENDER_CART_PATTERNS

    SKIP_RENDER_CART_PATTERNS.append('@@my_payment')


A ready-to-use implementation
=============================

Concrete implementations of the expected contracts already exists in
``bda.plone.shop`` for Archetypes and Dexterity. Please refer to
``https://github.com/bluedynamics/bda.plone.shop`` for more details.


Markup
======

Take a look at ``bda.plone.cart.browser:tile.pt`` how HTML markup
for adding items to cart might look like.

Basically a shop item consists of a container DOM element, containing an
element with CSS class ``cart_item_uid``, where the item UID is taken from:

.. code-block:: html

    <span class="cart_item_uid" style="display: none;">12345678</span>

a text input field with CSS class ``cart_item_count`` which is read for
item count:

.. code-block:: html

    <input type="text" size="2" value="1" class="cart_item_count" />

a quantity unit:

.. code-block:: html

    <span class="cart_item_quantity_unit">Quantity</span>

If quantity unit can be be float, add ``quantity_unit_float`` CSS class:

.. code-block:: html

    <input type="text" size="2" value="1"
           class="cart_item_count quantity_unit_float" />

the "add to Cart" action:

.. code-block:: html

    <a href="" class="add_cart_item">add to cart</a>

and the "update cart" action:

.. code-block:: html

    <a href="" class="update_cart_item">update cart</a>

Optionally, If cart viewlet is used, a status message can be displayed when
adding or updating cart items. This is useful if user should get
clearly informed if cart data has changed. To display status messages,
add CSS class ``show_status_message`` to "add to cart" and "update cart"
actions:

.. code-block:: html

    <a href="" class="update_cart_item show_status_message">update cart</a>

and optionally an element defining a comment or an input for entering a
comment:

.. code-block:: html

    <input type="text" size="16" value="" class="cart_item_comment" />

If comment should be required, add CSS class ``required`` to comment input.
If comment is empty, an error message gets shown to the user when trying to
add or update a cart item:

.. code-block:: html

    <input type="text" size="16" value="" class="cart_item_comment required" />


Create translations
===================

::

    $ cd src/bda/plone/cart/
    $ ./i18n.sh


Develop CSS Styles
==================

CSS Styles are written in Sass and compiled via `npm`::

    $ npm install
    $ npm run build  # to build the resources or
    $ npm run watch  # to watch changes during development


Contributors
============

- Robert Niederreiter (Autor)
- Sven Plage
- Peter Holzer
- Harald Friessnegger
- Peter Mathis
- Espen Moe-Nilssen
- Johannes Raggam
- Jure Cerjak
- Icons by famfamfam


CHANGES.rst

Changelog
=========

2.0.dev0 (unreleased)
---------------------

- rewrite LESS resources to SASS and remove grunt in favor of npm/sass
  [petschki]

- Use bs popover markup to display cart status message.
  [rnix]

- Remove decimal_converter and rely on plone.restapi>=6.13.7
  [agitator]

- Mark cart item in summary and portlet if comment is enabled or not.
  [jensens]

- Fix broken remove from cart in summary.
  [jensens]

- Handle count number localized.
  [jensens]

- No longer support for z3c.autoinclude.
  [jensens]

- More CSS classes on cart
  [jensens]

- Fix problem, when cart item value is removed while typing item was removed from cart.
  [jensens]

- Fix problem, when in listing the to be added number changes was immediately is added to the cart.
  [jensens]

- Moved shipping code over to here.
  [jensens]

- Refactored lots of code out of init into more semantic modules.
  [jensens]

- Added RESTAPI
  [jensens]

- Code style black.
  [jensens]

- Python 2/3 compatibility
  [agitator]

- Update version and classifiers - 2.x targets Plone 5.1/5.2 without Archetypes
  [agitator]

- Italian translations
  [ale-rt]

- switch to Plone 6 icon resolver
  [petschki]


1.0a1 (unreleased)
------------------

- Keep cart item cound focus after auto update.
  [rnix]

- Auto update cart if item count changes.
  [rnix]

- Replace unittest2 with untittest.
  [llisa123]

- Introduce ``bda.plone.cart.browser.portlet.SKIP_RENDER_CART_PATTERNS`` which
  can be used to define patterns which are checked against actual browser URL
  defining whether to skip cart rendering.
  [rnix]

- In ``cart.js``, make code blocks dealing with add and update after click event available for third party code.
  A move of the two anonymous functions in its own named functions,
  named ``update_cart_item`` and ``add_cart_item`` taking the node as parameter there were called on.
  [jensens]

- In ``cart.js``, allow also a select for count.
  [jensens. agitator]

- Add support for displaying cart item discount in cart.
  [rnix]

- Add ``get_item_shipping`` and ``cart_item_free_shipping`` utilities.
  [rnix]

- Fix typo. Rename remaining ``card_item_tempate`` to ``cart_item_tempate``.
  This already has been changed for summary and overview
  (in ``bda.plone.checkout``) but was missing in ``cart.js`` and ``tile.pt``
  fixes ``#31``. Make sure to fix this in all your cart related custom
  templates!
  [rnix]

- More precise selectors for cart item template, cart summary discount and
  cart summary shipping in ``cart.js``.
  [rnix]

- Use code for safe uid also for get_object_by_uid and use ``plone.api``.
  [jensens]

- fixed selector typo #cart_viewlet_summery with #cart_viewlet_summary
  [agitator]

- Pass comment in JS from add/update cart to server side validation.
  [jensens]

- use MwSt. instead of Ust. in german
  [agitator]

- added data-context-url for sane cartData and validate_cart_item calls on Plone 5
  [agitator]

- Plone 5 update
  [agitator]

- ``extractitems`` now returns a namedtuple with attributes ``uid``, ``count`` and ``comment``
  [jensens]


0.10.dev2
---------

- Resolve JSHint errors and warnings.
  [thet]

- Fix cart item not removable, if comment is required. Fixes #17, fixes #12.
  [thet]

- Remove portlet's cartWrapper div node. It's not used at all and only
  introduces non-standard portlet HTML markup.
  [thet]

- By default, cart viewlet is hidden and gets displayed as soon as cart items
  are fetched.
  [rnix]

- Consider ``hide cart if empty`` setting on cart viewlet.
  [rnix]

- Hide unset preview images for cart item.
  [rnix]

- Cart can display status message if cart data gets modified.
  [rnix]

- Cart might be rendered as viewlet in portal header instead of as portlet.
  [rnix]

- Do not expect ``IShippingItem`` to be implemented. See also
  ``https://github.com/bluedynamics/bda.plone.shop/issues/31``
  [jensens]

- No caching/merging of cart.translations.js
  [agitator]



0.9
---

- Quote/Unqote special characters in cart cookie only for cart item comment.
  Thus we can have any characters in comment. "Invalid comment characters"
  validation and error message no longer necessary.
  [rnix]


0.8
---

- Use ``decodeURIComponent`` instead of deprectaed ``unescape`` in ``cart.js``.
  [rnix]

- Catch ``ValueError`` if given uid value is no valid ``uuid.UUID`` string in
  ``get_catalog_brain``.
  [rnix]

- Let ``get_catalog_brain`` and thus ``get_object_by_uid`` handle ``uuid.UUID``
  objects, hex strings like '8de81513431752d5f32c680db93dda0c' and UUID object
  representation strings like '8de81513-4317-52d5-f32c-680db93dda0c'.
  [thet]

- Encode umlaut characters in cart item comment.
  [rnix]


0.7
---

- Add comment label in cart summary template.
  [rnix]


0.6
---

- Hide discount and shipping info in cart via JS if data changed respective.
  [rnix]

- Remove ``include_shipping_costs`` property from ``CartView``.
  [rnix]

- Always deliver shipping markup for cart, control with JS whether shipping
  costs are displayed.
  [rnix]

- Move ``bda.plone.cart.browser.CURRENCY_LITERALS`` to
  ``bda.plone.cart.CURRENCY_LITERALS``.
  [rnix]

- Add ``ICartDataProvider.max_artice_count`` attribute and implement in
  ``CartDataProviderBase``.
  [rnix]

- Add ``bda.plone.cart.cart_item_shippable`` utility.
  [rnix]

- Implement ``ICartDataProvider.include_shipping_costs`` on
  ``CartDataProviderBase`` using
  ``bda.plone.shipping.interfaces.IShippingItem.shipping`` flag for
  calculation.
  [rnix]

- Add ``ICartDataProvider.total`` attribute and implement in
  ``CartDataProviderBase``.
  [rnix]

- Use ``readcookie`` instead of expecting ``items`` request parameter in
  ``CartDataProviderBase.data``.
  [rnix]

- Do not pass recent cart ``items`` parameter to ``@@cartData`` view, contained
  value is included as ``cart`` cookie anyways.
  [rnix]

- Adopt shipping handling to ``bda.plone.shipping`` >= 0.4.
  [rnix]


src/bda/plone/cart/cart.py
# -*- coding: utf-8 -*-
from bda.plone.cart import _
from bda.plone.cart import cartitem
from bda.plone.cart import cookie
from bda.plone.cart import utils
from bda.plone.cart.interfaces import ICartDataProvider
from bda.plone.cart.interfaces import ICartDiscount
from bda.plone.cart.shipping import Shippings
from decimal import Decimal
from zope.component import adapter
from zope.component import getMultiAdapter
from zope.component import queryAdapter
from zope.i18n import translate
from zope.interface import implementer
from zope.interface import Interface
from zope.publisher.interfaces import IRequest


def get_data_provider(context, request=None):
    """Return ICartDataProvider implementation.
    """
    if request is None:
        request = context.REQUEST
    return getMultiAdapter((context, request), ICartDataProvider)


@implementer(ICartDataProvider)
@adapter(Interface, IRequest)
class CartDataProviderBase(object):
    def __init__(self, context, request):
        self.context = context
        self.request = request

    @property
    def data(self):
        ret = dict()
        ret["cart_settings"] = dict()
        ret["cart_settings"]["hide_cart_if_empty"] = self.hide_cart_if_empty
        if self.disable_max_article:
            ret["cart_settings"]["cart_max_article_count"] = 10000
        else:
            ret["cart_settings"]["cart_max_article_count"] = self.max_artice_count
        include_shipping_costs = self.include_shipping_costs
        ret["cart_settings"]["include_shipping_costs"] = include_shipping_costs
        ret["cart_items"] = list()
        ret["cart_summary"] = dict()
        items = cookie.extract_items(cookie.read(self.request))
        if items:
            net = self.net(items)
            vat = self.vat(items)
            ret["cart_items"] = self.cart_items(items)
            ret["cart_summary"]["cart_vat"] = utils.ascur(vat)
            ret["cart_summary"]["cart_net"] = utils.ascur(net)
            cart_discount = self.discount(items)
            discount_net = cart_discount["net"]
            discount_vat = cart_discount["vat"]
            discount_total = discount_net + discount_vat
            ret["cart_summary"]["discount_net"] = "-" + utils.ascur(discount_net)
            ret["cart_summary"]["discount_vat"] = "-" + utils.ascur(discount_vat)
            ret["cart_summary"]["discount_total"] = "-" + utils.ascur(discount_total)
            ret["cart_summary"]["discount_total_raw"] = discount_total
            total = net + vat - discount_total
            if include_shipping_costs:
                shipping = self.shipping(items)
                total += shipping["net"] + shipping["vat"]
                label = translate(shipping["label"], context=self.request)
                ret["cart_summary"]["shipping_label"] = label
                if shipping["description"]:
                    desc = translate(shipping["description"], context=self.request)
                    ret["cart_summary"]["shipping_description"] = desc
                else:
                    ret["cart_summary"]["shipping_description"] = ""
                ret["cart_summary"]["shipping_net"] = utils.ascur(shipping["net"])
                ret["cart_summary"]["shipping_vat"] = utils.ascur(shipping["vat"])
                ret["cart_summary"]["shipping_total"] = utils.ascur(
                    shipping["net"] + shipping["vat"]
                )
                ret["cart_summary"]["shipping_total_raw"] = (
                    shipping["net"] + shipping["vat"]
                )
                # B/C for bda.plone.cart < 0.6 custom templates
                ret["cart_summary"]["cart_shipping"] = utils.ascur(
                    shipping["net"] + shipping["vat"]
                )
            ret["cart_summary"]["cart_total"] = utils.ascur(total)
            ret["cart_summary"]["cart_total_raw"] = total
        return ret

    @property
    def total(self):
        total = Decimal(0)
        items = cookie.extract_items(cookie.read(self.request))
        net = self.net(items)
        vat = self.vat(items)
        cart_discount = self.discount(items)
        discount_net = cart_discount["net"]
        discount_vat = cart_discount["vat"]
        discount_total = discount_net + discount_vat
        total = net + vat - discount_total
        if self.include_shipping_costs:
            shipping = self.shipping(items)
            total += shipping["net"] + shipping["vat"]
        return total.quantize(Decimal("1.000"))

    @property
    def currency(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``currency``."
        )

    @property
    def hide_cart_if_empty(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``hide_cart_if_empty``."
        )

    @property
    def max_artice_count(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``max_artice_count``."
        )

    @property
    def disable_max_article(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``disable_max_article``."
        )

    @property
    def summary_total_only(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``summary_total_only``."
        )

    @property
    def include_shipping_costs(self):
        items = cookie.extract_items(cookie.read(self.request))
        for item in items:
            if cartitem.cart_item_shippable(self.context, item):
                return True
        return False

    @property
    def shipping_method(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``shipping_method``."
        )

    @property
    def checkout_url(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``checkout_url``."
        )

    @property
    def cart_url(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``cart_url``."
        )

    @property
    def show_to_cart(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``show_to_cart``."
        )

    @property
    def show_checkout(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``show_checkout``."
        )

    @property
    def show_currency(self):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``show_currency``."
        )

    def validate_set(self, uid):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``validate_set``."
        )

    def validate_count(self, uid, count):
        """Validate setting cart item count for uid.

        uid - Is the cart item UID.
        count - If count is 0, it means that a cart item is going to be
        deleted, which is always allowed. If count is > 0, it's the aggregated
        item count in cart.
        """
        cart_item = utils.get_object_by_uid(self.context, uid)
        item_data = cartitem.get_item_data_provider(cart_item)
        cart_count_limit = item_data.cart_count_limit
        if cart_count_limit and float(count) > cart_count_limit:
            message = translate(
                _("article_limit_reached", default="Article limit reached"),
                context=self.request,
            )
            return {"success": False, "error": message, "update": False}
        item_state = cartitem.get_item_state(cart_item, self.request)
        if item_state.validate_count(count):
            return {"success": True, "error": ""}
        message = translate(
            _(
                "trying_to_add_more_items_than_available",
                default="Not enough items available, abort.",
            ),
            context=self.request,
        )
        return {"success": False, "error": message, "update": False}

    def net(self, items):
        raise NotImplementedError(u"CartDataProviderBase does not implement ``net``.")

    def vat(self, items):
        raise NotImplementedError(u"CartDataProviderBase does not implement ``vat``.")

    def shipping(self, items):
        shippings = Shippings(self.context)
        shipping = shippings.get(self.shipping_method)
        try:
            return {
                "label": shipping.label,
                "description": shipping.description,
                "net": shipping.net(items),
                "vat": shipping.vat(items),
            }
        # B/C for bda.plone.shipping < 0.4
        except NotImplementedError:
            return {
                "label": shipping.label,
                "description": shipping.description,
                "net": shipping.calculate(items),
                "vat": Decimal(0),
            }

    def discount(self, items):
        net = vat = Decimal(0)
        discount = queryAdapter(self.context, ICartDiscount)
        if discount:
            net = discount.net(items)
            vat = discount.vat(items)
        return {"net": net, "vat": vat}

    def cart_items(self, items):
        raise NotImplementedError(
            u"CartDataProviderBase does not implement ``cart_items``."
        )

    def item(
        self,
        uid,
        title,
        count,
        price,
        url,
        comment="",
        description="",
        comment_enabled=False,
        comment_required=False,
        quantity_unit_float=False,
        quantity_unit="",
        preview_image_url="",
        no_longer_available=False,
        alert="",
        discount=Decimal(0),
    ):
        return {
            # placeholders
            "cart_item_uid": uid,
            "cart_item_title": title,
            "cart_item_count": count,
            "cart_item_price": utils.ascur(price),
            "cart_item_location:href": url,
            "cart_item_preview_image:src": preview_image_url,
            "cart_item_comment": comment,
            "cart_item_description": description,
            "cart_item_quantity_unit": quantity_unit,
            "cart_item_alert": alert,
            "cart_item_discount": utils.ascur(discount)
            if discount != Decimal(0)
            else Decimal(0),
            # control flags
            "comment_enabled": comment_enabled,
            "comment_required": comment_required,
            "quantity_unit_float": quantity_unit_float,
            "no_longer_available": no_longer_available,
        }
